validationinterface: make MainSignalsInstance() a class, drop unused forward declarations

jonatack opened the pull request:
Make MainSignalsInstance a class, rename it to MainSignalsImpl, use Doxygen documentation for it, and remove no longer used forward declarations in src/validationinterface.h.

----
    
MainSignalsInstance was created in 3a19fed9db5 and originally was a collection of boost::signals methods moved to validationinterface.cpp, in order to no longer need to include boost/signals in validationinterface.h.

MainSignalsInstance then evolved in d6815a23131 to become class-like:
    
- [C.8: Use class rather than struct if any member is non-public](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-class)
    
 - [C.2: Use class if the class has an invariant; use struct if the data members can vary independently](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c2-use-class-if-the-class-has-an-invariant-use-struct-if-the-data-members-can-vary-independently)
   
- A class has the advantage of default private access, as opposed to public for a struct.

jonatack committed:
refactor: make MainSignalsInstance() a class

and use Doxygen documentation for it, per our developer notes.

Context:

MainSignalsInstance was created in 3a19fed9db5 and originally was a struct
collection of boost::signals methods moved to validationinterface.cpp, in order
to no longer need to include boost/signals in validationinterface.h.

MainSignalsInstance then evolved in d6815a23131 to remove boost/signals2 and became class-like.

[C.8: Use class rather than struct if any member is
non-public](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-class)

[C.2: Use class if the class has an invariant; use struct if the data members can vary
independently](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c2-use-class-if-the-class-has-an-invariant-use-struct-if-the-data-members-can-vary-independently)

A class also has the advantage of default private access, as opposed to public for a struct.

jonatack commented:
Most of the diff is move-only and is easily reviewed with git options like --ignore-all-space -U8 --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space.

pk-b2 commented on src/validationinterface.h:
```diff
@@ -177,7 +179,43 @@ class CValidationInterface {
+ * callbacks.
+ *
+ * A std::unordered_map is used to track what callbacks are currently
+ * registered, and a std::list is to used to store the callbacks that are
```
nit:

```suggestion
 * registered, and a std::list is used to store the callbacks that are
```

jonatack replied:
Thanks, fixed.

pk-b2 reviewed (approved):
ACK 1949586afc2f57061d27b87bf48d40861b8a4767

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/25067/commits/2911d9460b6837653594cb65b6e4f3bd7fbb7e80

pk-b2 commented:
ACK https://github.com/bitcoin/bitcoin/pull/25067/commits/2911d9460b6837653594cb65b6e4f3bd7fbb7e80

MarcoFalke commented:
Is there a reason to move it to the header? This will make the header heavier and increase the includes, both of which theoretically result in slower builds.

jonatack committed:
refactor: remove unused forward declarations in validationinterface.h

MarcoFalke commented on src/validationinterface.h:
```diff
@@ -17,9 +18,7 @@ class BlockValidationState;
 struct CBlockLocator;
-class CConnman;
 class CValidationInterface;
-class uint256;
```
For reference, with

```diff
diff --git a/ci/test/06_script_b.sh b/ci/test/06_script_b.sh
index e64af2ad5..58661ff4c 100755
--- a/ci/test/06_script_b.sh
+++ b/ci/test/06_script_b.sh
@@ -35,14 +35,13 @@ if [ "$RUN_FUNCTIONAL_TESTS" = "true" ]; then
 fi
 
 if [ "${RUN_TIDY}" = "true" ]; then
-  export P_CI_DIR="${BASE_BUILD_DIR}/bitcoin-$HOST/src/"
-  CI_EXEC run-clang-tidy "${MAKEJOBS}"
   export P_CI_DIR="${BASE_BUILD_DIR}/bitcoin-$HOST/"
   CI_EXEC "python3 ${DIR_IWYU}/include-what-you-use/iwyu_tool.py"\
           " src/compat"\
           " src/init"\
           " src/rpc/fees.cpp"\
           " src/rpc/signmessage.cpp"\
+          " src/validationinterface.cpp"\
           " -p . ${MAKEJOBS} -- -Xiwyu --cxx17ns -Xiwyu --mapping_file=${BASE_BUILD_DIR}/bitcoin-$HOST/contrib/devtools/iwyu/bitcoin.core.imp"
 fi
 
```


I get:

```
validationinterface.h should add these lines:
#include <stddef.h>                  // for size_t
#include <stdint.h>                  // for uint64_t
#include "threadsafety.h"            // for LOCKS_EXCLUDED

validationinterface.h should remove these lines:
- class CConnman;  // lines 20-20
- class uint256;  // lines 22-22

validationinterface.cpp should add these lines:
#include <assert.h>                  // for assert
#include <ext/alloc_traits.h>        // for __alloc_traits<>::value_type
#include <iterator>                  // for next
#include <list>                      // for list<>::iterator, _List_iterator
#include <vector>                    // for vector
#include "attributes.h"              // for LIFETIMEBOUND
#include "sync.h"                    // for AssertLockNotHeldInternal, LOCK
#include "uint256.h"                 // for uint256

jonatack replied:
Thanks. Dropped the header changes.

maflcko reviewed (commented):
second commit looks wrong:

* "enter the commit message for your changes. Lines starting"
* missing includes (see iwyu)

jonatack commented:
> Not sure about fixing includes manually now that we have iwyu. Seems wasteful of review resources and is likely still wrong (as seen here).

Yes, I hesitated on it and agree. Dropped the headers.

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/25067/commits/c05ee1e8440f7e1030a5d9597cb6963b99fe4051

promag reviewed (commented):
Code review ACK c05ee1e8440f7e1030a5d9597cb6963b99fe4051.

nit, would use `refactor:` prefix instead, like:

 - refactor: make MainSignalsInstance a class
 - refactor: remove unused forward declarations

2nd nit, I wouldn't mind dropping `Instance` from `MainSignalsInstance`. Maybe rename it to `MainSignalsImpl` as it looks like `CMainSignals` follows pimpl.

jonatack committed:
scripted-diff: Rename MainSignalsInstance() class to MainSignalsImpl()

```
-BEGIN VERIFY SCRIPT-
s() { git grep -l "$1" src test doc | xargs sed -i "s/$1/$2/g"; }
s 'MainSignalsInstance' 'MainSignalsImpl'
-END VERIFY SCRIPT-

jonatack commented:
> nit, would use `refactor:` prefix instead, like:
> 
> 2nd nit, I wouldn't mind dropping `Instance` from `MainSignalsInstance`. Maybe rename it to `MainSignalsImpl` as it looks like `CMainSignals` follows pimpl.

Done. Grepping for `MakeSignals` would yield many other results, so renamed to `MainSignalsImpl` as proposed.

promag reviewed (commented):
Code review ACK ca1ac1f0e0fbabbe97666aca94024afb8104da06.

danielabrozzoni reviewed (approved):
Code review ACK ca1ac1f0e0fbabbe97666aca94024afb8104da06

furszy commented:
Code ACK ca1ac1f

Tiny nit for the renaming commit ca1ac1f: 
Isn't the "main" word in `MainSignalsImpl` coming from the previous monolithic `main.cpp` file existence?. Probably could be changed for something else as now the `MainSignalsImpl` is just a generic container class for the "validation interface" events subscribers.

jonatack commented:
> Isn't the "main" word in `MainSignalsImpl` coming from the previous monolithic `main.cpp` file existence?

Hm, perhaps. I've been reading it to mean "main signals" as in "primary signals" and there are also CMainSignals and GetMainSignals in validationinterface.{h,cpp}, so I don't have a strong opinion.
